### PR TITLE
ENH: _dir_additions returns also the first level of a MultiIndex

### PR DESCRIPTION
**Pull request by BibMartin**
- [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

Assuming a DataFrame with MultiIndex columns like this:
```python
>> df
    foo   bar      
          foo   bar
0  0.32  0.15  0.45
1  0.62  0.73  0.36
2  0.21  0.68  0.10
3  0.05  0.36  0.90
```

one can access to `df['foo']` and `df['bar']` with the shortcuts `df.foo` and `df.bar`,
but one don't benefit from autocompletion because 'foo' and 'bar' are not listed in `dir(df)`. 

```python
>> 'bar' in dir(df)
False
```

This PR extends `df._dir_additions` so that the first level of a MultiIndex is listed in `dir(df)`.

```python
>> 'bar' in dir(df)
True
```

### CODE REVIEW

**Comment by jreback** [comment]
I am +0 on this. This add some needless complexity, though from a user POV might be nice. Would need some additional tests to consider.

**Comment by BibMartin** [comment]
@jreback 
> Would need some additional tests to consider.

Sure. I plan to put tests if the principle is accepted. Any idea is welcome.
Btw, I've not seen whether there are tests about `dir` behavior.

**Comment by jreback** [comment]
> Btw, I've not seen whether there are tests about dir behavior.

``bash-3.2$ grep -R 'dir(' pandas/tests/``

**Comment by BibMartin** [comment]
Added `test_tab_completion` for DataFrames, inpired from Series.

**Comment by jreback** [comment]
@jorisvandenbossche thoughts here

**Comment by jreback** [comment]
this is ok in principle. can you rebase / update

**Comment by jreback** [comment]
can you rebase

**Comment by BibMartin** [comment]
Sorry, I've been long to rebase.

Do I need to add a whatsneww entry ? In which version file ?

File: pandas/core/generic.py [hunk 1]
```diff
@@ -196,7 +196,11 @@ def __unicode__(self):
     def _dir_additions(self):
         """ add the string-like attributes from the info_axis """
         additions = set([c for c in self._info_axis
-                         if isinstance(c, string_types) and isidentifier(c)])
+                         if isinstance(c, string_types) and isidentifier(c)] +
+                        [c[0] for c in self._info_axis
```
Comment (jreback) [inline pandas/core/generic.py]: I would rather both of these to (private) methods on ``Index`` and overriden on a ``MultiIndex``; this become much simpler

e.g.
```
def _to_identifiers(self):
    return [c for c in self if isinstance(c, string_types) and isidentifier(c)]
```

with some doc-strings & tests

then this method becomes really simple.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (BibMartin) [reply pandas/core/generic.py]: Now that I understand the code better, I realize that another simple implementation could be:
```python
additions = set([c for c in self._info_axis.get_level_values(0)
                 if isinstance(c, string_types) and isidentifier(c)])
```
Please tell me which method would be best.

Comment (jreback) [reply pandas/core/generic.py]: yes this looks reasonable. pls add a comment on what is going one.

**Comment by BibMartin** [comment]
Changes are done, and rebased. Btw, I've added a test on Series auto-completion.

File: pandas/tests/series/test_api.py [hunk 2]
```diff
@@ -234,6 +234,39 @@ def test_tab_completion(self):
         assert 'str' not in dir(s)
         assert 'dt' in dir(s)  # as it is a datetime categorical
 
+    def test_index_tab_completion(self):
+        # dir contains string-like values of the Index.
```
Comment (jreback) [inline pandas/tests/series/test_api.py]: you can parametrize this with index

File: pandas/tests/series/test_api.py [hunk 3]
```diff
@@ -234,6 +234,39 @@ def test_tab_completion(self):
...
+                      tm.makeFloatIndex(100),
+                      Index([True, False]),
+                      ]:
+            s = pd.Series(index=index)
+            dir_s = dir(s)
+            for x in s.index:
+                assert (not isinstance(x, string_types) or
+                        not isidentifier(x) or x in dir_s)
+
+        # dir contains string-like values of the MultiIndex first level.
```
Comment (jreback) [inline pandas/tests/series/test_api.py]: you can add these in the parameterization above

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a note in 0.22, other enhancements.

**Comment by jreback** [comment]
one additional thing I think we need. If you have a very large index, ``_dir_additions`` actually takes quite a bit of time (this is not exclusive of this change).

So what I would do is if the index is say < 100, use the currently ``_dir_addition``, otherwise return an empty list! (its essentially too big to use tab completion for anyhow). can you make this change and add an asv for this (could be a separate PR as well)

**Comment by jreback** [comment]
you can actually use ``.unique(level=0)`` to make this efficient

File: doc/source/whatsnew/v0.22.0.txt [hunk 4]
```diff
@@ -77,6 +77,8 @@ Other Enhancements
 - :func:`Series.fillna` now accepts a Series or a dict as a ``value`` for a categorical dtype (:issue:`17033`)
 - :func:`pandas.read_clipboard` updated to use qtpy, falling back to PyQt5 and then PyQt4, adding compatibility with Python3 and multiple python-qt bindings (:issue:`17722`)
 - Improved wording of ``ValueError`` raised in :func:`read_csv` when the ``usecols`` argument cannot match all columns. (:issue:`17301`)
+- :func:`NDFrame._dir_additions` (tab completion) also returns identifiers in the first level of a :func:`MultiIndex`. (:issue:`16326`)
+- :func:`NDFrame._dir_additions` (tab completion) limits to 100 values, for better performance. (:issue:`18587`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: can you move 2nd to performance. don't use NDFrame. just say Series/DataFrame tab completion

File: pandas/tests/frame/test_api.py [hunk 5]
```diff
@@ -128,6 +128,20 @@ def test_column_contains_typeerror(self):
         except TypeError:
             pass
 
+    def test_tab_completion(self):
+        # DataFrame whose columns are identifiers shall have them in __dir__.
+        df = pd.DataFrame([list('abcd'), list('efgh')], columns=list('ABCD'))
+        assert 'A' in dir(df)
```
Comment (jreback) [inline pandas/tests/frame/test_api.py]: actually you can assert all of the items are in the dir() (use a loop)

File: pandas/tests/frame/test_api.py [hunk 6]
```diff
@@ -128,6 +128,20 @@ def test_column_contains_typeerror(self):
...
+        assert isinstance(df.__getitem__('A'), pd.Series)
+
+        # DataFrame whose first-level columns are identifiers shall have
+        # them in __dir__.
+        df = pd.DataFrame(
+            [list('abcd'), list('efgh')],
+            columns=pd.MultiIndex.from_tuples(list(zip('ABCD', 'EFGH'))))
+        assert 'A' in dir(df)
+        assert isinstance(df.__getitem__('A'), pd.DataFrame)
+
```
Comment (jreback) [inline pandas/tests/frame/test_api.py]: assert EFGH are NOT in the result (in a loop) and assert ABCD there

File: pandas/tests/series/test_api.py [hunk 7]
```diff
@@ -10,7 +10,7 @@
 from pandas import Index, Series, DataFrame, date_range
 from pandas.core.indexes.datetimes import Timestamp
 
-from pandas.compat import range
+from pandas.compat import range, lzip, isidentifier, string_types
```
Comment (jreback) [inline pandas/tests/series/test_api.py]: can you add a test that has 101 columns and assert first 100 there and last 1 is not.

File: pandas/core/generic.py [hunk 8]
```diff
@@ -195,9 +195,12 @@ def __unicode__(self):
         return '%s(%s)' % (self.__class__.__name__, prepr)
 
     def _dir_additions(self):
-        """ add the string-like attributes from the info_axis """
-        additions = set([c for c in self._info_axis
-                         if isinstance(c, string_types) and isidentifier(c)])
+        """ add the string-like attributes from the info_axis.
+        If info_axis is a MultiIndex, it's first level values are used.
+        """
+        additions = set(
```
Comment (jreback) [inline pandas/core/generic.py]: can you add an asv that does dir() on a Series/DataFrame with say 10000 elements (we *might* already have one of these)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by BibMartin** [comment]
Rebased and ready

File: pandas/tests/series/test_api.py (hunk 7 above)
Comment (jreback) [reply pandas/tests/series/test_api.py]: can you do thsi

Comment (BibMartin) [reply pandas/tests/series/test_api.py]: I've put this in the other test (See [here](https://github.com/pandas-dev/pandas/pull/16326/files#diff-db91df05b120c5eb39d90899b54b5321R266) and [there](https://github.com/pandas-dev/pandas/pull/16326/files#diff-db91df05b120c5eb39d90899b54b5321R277))
Please tell me if you prefer that I put it in a separate test.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @BibMartin nice patch!

**Comment by BibMartin** [comment]
Thanks to you @jreback : it's been a great occasion to learn for me.

**Pull request merged**